src/vanilla.ts

```diff
@@ -56,11 +56,16 @@ export type Snapshot<T> = T extends { $$valtioSnapshot: infer S }
 
 type RemoveListener = () => void
 type AddListener = (listener: Listener) => RemoveListener
+type AddPropListener = (
+  prop: string | number | symbol,
+  listener: Listener,
+) => RemoveListener
 
 type ProxyState = readonly [
   target: object,
   ensureVersion: (nextCheckVersion?: number) => number,
   addListener: AddListener,
+  addPropListener: AddPropListener,
```

Comment:
Why can't you use addListener and check prop value / version in your callback function? I believe it can still accomplish O(1)-useSnapshot. (My goal is to allow your third-party library to work without changing vanilla.ts without changing much.)

Reply:
Yes. `addPropListener` logic can move outside vanilla.ts, but I found `createSnapshot` still need to modify, so I just left `addPropListener` there until I got full solution in my mind.

Reply:
Do you modify createSnapshot in vanilla.ts?
I'm not sure what's blocking for your O(1)-useSnapshot without modifying vanilla.ts. I hope you find a solution. If not, you can try as much as possible, and I hope we can find something from there.

Reply:
So, is it only about getter? Do you mean it works if we only use plain objects? Then, it sounds like a different requirement from O(1)-useSnapshot.

Reply:
Most likely, yes. And some edge case about ProxySet and ProxyMap. I havn't give it a try. 
Also there are some dependencies on `proxy-compare` in `vanilla.ts`, I replace it with my self version, since I can't get the interval variables in `proxy-compare`.

Reply:
At this point, I'm fairly confident that a third-party library can implement O(1)-useSnapshot, maybe without proxy-compare.

Yeah, I agree that it's not super cool that vanilla.ts depends on proxy-compare. I'll see how it will be a breaking change if we drop it.